apps/dotcom/src/components/DocumentName/DocumentName.tsx

```diff
@@ -0,0 +1,287 @@
+		element.style.setProperty('max-width', maxWidth + 'px')
+	}, [])
+
+	useLayoutEffect(() => {
+		const element = ref.current
+		if (!element) return
+
+		const layoutTop = element.parentElement!.parentElement!
+		const leftPanel = layoutTop.querySelector('.tlui-layout__top__left')! as HTMLElement
+		const rightPanel = layoutTop.querySelector('.tlui-layout__top__right')! as HTMLElement
+
+		// Update layout when the things change
+		const observer = new ResizeObserver(updateLayout)
+		observer.observe(leftPanel)
+		observer.observe(rightPanel)
+		observer.observe(layoutTop)
+		observer.observe(element)
+
+		// Also update on first layout
+		updateLayout()
+
+		return () => {
+			observer.disconnect()
+		}
+	}, [updateLayout])
+
+	// Update after every render, too
+	useLayoutEffect(() => {
+		updateLayout()
+	})
+
+	return (
+		<div ref={ref} className="tlui-top-zone__container">
+			{children}
+		</div>
+	)
+}
+
+const DocumentNameEditor = track(function DocumentNameEditor({
+	state,
+	setState,
+}: {
+	state: NameState
+	setState: (update: SetStateAction<NameState>) => void
+}) {
+	const inputRef = useRef<HTMLInputElement>(null)
+	const editor = useEditor()
+	const documentSettings = editor.getDocumentSettings()
+	const msg = useTranslation()
+	const defaultDocumentName = msg('document.default-name')
+
+	useEffect(() => {
+		if (state.isEditing && inputRef.current) {
+			inputRef.current.select()
+		}
+	}, [state.isEditing])
+
+	useEffect(() => {
+		const save = () => {
+			if (state.name === null) return
+			const trimmed = state.name.trim()
+			if (!trimmed || trimmed === documentSettings.name.trim()) {
+				if (!state.isEditing) setState((prev) => ({ ...prev, name: null }))
+				return
+			}
+
+			editor.updateDocumentSettings({ name: trimmed })
+		}
+
+		if (!state.isEditing) {
+			save()
+		}
+	}, [documentSettings.name, editor, state.isEditing, state.name, setState])
+
+	const handleChange = useCallback(
+		(e: ChangeEvent<HTMLInputElement>) => {
+			const value = e.currentTarget.value
+			setState((prev) => ({ ...prev, name: value }))
+		},
+		[setState]
+	)
+
+	const handleKeydownCapture = useCallback(
+		(e: KeyboardEvent) => {
+			if (e.key === 'Enter') {
+				e.preventDefault()
+				// blur triggers save
+				inputRef.current?.blur()
+			} else if (e.key === 'Escape') {
+				e.preventDefault()
+				// revert to original name instantly so that when we blur we don't
+				// trigger a save with the new one
+				setState((prev) => ({ ...prev, name: null }))
+				inputRef.current?.blur()
+			}
+		},
+		[setState]
+	)
+
+	const handleBlur = useCallback(() => {
+		setState((prev) => ({ ...prev, isEditing: false }))
+	}, [setState])
+
+	const name = state.name || documentSettings.name || defaultDocumentName
+
+	return (
+		<div className="tlui-document-name__input__wrapper">
+			{state.isEditing && (
+				<input
+					ref={inputRef}
+					className="tlui-document-name__input"
+					value={state.name ?? documentSettings.name}
```

Comment:
you're comparing to `state.name` to `null` in other places so i think this logic here won't work how you want it too. prbly you want `||` instead of `??`

Reply:
Looks like this one is correct, with an `||` it would not allow us to clear the input:

https://github.com/tldraw/tldraw/assets/2523721/1b5fe3e8-93ed-4114-81fa-8dc0f30942e6